examples/blake2f.rs

```diff
@@ -0,0 +1,1459 @@
+                    + bit_values[i * 4 + 2] * 4
+                    + bit_values[i * 4 + 3] * 8,
+            )
+        })
+        .collect()
+}
+
+fn g_wg<F: PrimeField + Hash>(
+    (v1_vec_values, v2_vec_values): (&mut [u64], &mut [u64]),
+    (a, b, c, d): (usize, usize, usize, usize),
+    (x, y): (u64, u64),
+    (va_bit_vec, vb_bit_vec): (&mut Vec<Vec<F>>, &mut Vec<Vec<F>>),
+    (vc_bit_vec, vd_bit_vec): (&mut Vec<Vec<F>>, &mut Vec<Vec<F>>),
+    (v_xor_d_bit_vec, v_xor_b_bit_vec): (&mut Vec<Vec<F>>, &mut Vec<Vec<F>>),
+    (b_bit_vec, b_3bits_vec): (&mut Vec<F>, &mut Vec<F>),
+) {
+    va_bit_vec.push(split_value_4bits(
+        v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + x as u128,
+        BITS_NUMBER + 1,
+    ));
+    v1_vec_values[a] = (v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + x as u128) as u64;
+
+    vd_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+    v1_vec_values[d] = ((v1_vec_values[d] ^ v1_vec_values[a]) >> R1)
+        ^ (v1_vec_values[d] ^ v1_vec_values[a]) << (64 - R1);
+    v_xor_d_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+
+    vc_bit_vec.push(split_value_4bits(
+        v1_vec_values[c] as u128 + v1_vec_values[d] as u128,
+        BITS_NUMBER + 1,
+    ));
+    v1_vec_values[c] = (v1_vec_values[c] as u128 + v1_vec_values[d] as u128) as u64;
+
+    vb_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+    v1_vec_values[b] = ((v1_vec_values[b] ^ v1_vec_values[c]) >> R2)
+        ^ (v1_vec_values[b] ^ v1_vec_values[c]) << (64 - R2);
+    v_xor_b_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+
+    va_bit_vec.push(split_value_4bits(
+        v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + y as u128,
+        BITS_NUMBER + 1,
+    ));
+    v2_vec_values[a] = (v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + y as u128) as u64;
+
+    vd_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+    v2_vec_values[d] = ((v1_vec_values[d] ^ v2_vec_values[a]) >> R3)
+        ^ (v1_vec_values[d] ^ v2_vec_values[a]) << (64 - R3);
+    v_xor_d_bit_vec.push(split_value_4bits(v2_vec_values[d] as u128, BITS_NUMBER));
+
+    vc_bit_vec.push(split_value_4bits(
+        v1_vec_values[c] as u128 + v2_vec_values[d] as u128,
+        BITS_NUMBER + 1,
+    ));
+    v2_vec_values[c] = (v1_vec_values[c] as u128 + v2_vec_values[d] as u128) as u64;
+
+    vb_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+    v2_vec_values[b] = ((v1_vec_values[b] ^ v2_vec_values[c]) >> R4)
+        ^ (v1_vec_values[b] ^ v2_vec_values[c]) << (64 - R4);
+    v_xor_b_bit_vec.push(split_value_4bits(
+        (v1_vec_values[b] ^ v2_vec_values[c]) as u128,
+        BITS_NUMBER,
+    ));
+    let bits = (v1_vec_values[b] ^ v2_vec_values[c]) / 2u64.pow(60);
+    b_bit_vec.push(F::from(bits / 8));
+    b_3bits_vec.push(F::from(bits % 8))
+}
+
+fn g_setup<F: PrimeField + Hash>(
```

Comment:
@10to4 we need to explain in a comment that one call to g_setup just constraint half of what the G function in the spec does.

Reply:
ok